File: GFG/Stacks/nxtgreat.cpp
#include<iostream>
#include"../ds.h"
using namespace std;

void nxtgrt(vector<int> arr){
    int n=arr.size();
    int i=n-2;
    Stack<int> s;
    s.push(arr[n-1]);
    cout<<"-1"<<" ";
    while(i>=0){
        while(s.isEmpty()==false && s.peek()<arr[i]){
            s.pop();
        }
        (s.isEmpty()) ? cout<<"-1"<<" " : cout<<s.peek()<<" ";
        s.push(arr[i]);
        i--;
    }
}

int main(){
    vector<int> arr {15,10,18,12,4,6,2,8};
    nxtgrt(arr);
    return 0;
}

File: GFG/Stacks/nxtgreat.py
import ds

def nxtgrt(arr):
    n=len(arr)
    i=n-2
    s=ds.Stack()
    s.push(arr[-1])
    print("-1",end=" ")
    while(i>=0):
        while(s.isEmpty()==False and s.top()<arr[i]):
            s.pop()
        print("-1",end=" ") if(s.isEmpty()) else (print(s.top(),end=" "))
        s.push(arr[i])
        i-=1
arr=[15,10,18,12,4,6,2,8]
nxtgrt(arr)